ext/standard/streamsfuncs.c

```diff
@@ -756,116 +809,141 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
 /* }}} */
 
 /* {{{ Runs the select() system call on the sets of streams with a timeout specified by tv_sec and tv_usec */
-PHP_FUNCTION(stream_select)
-{
-	zval *r_array, *w_array, *e_array;
-	struct timeval tv, *tv_p = NULL;
-	fd_set rfds, wfds, efds;
-	php_socket_t max_fd = 0;
-	int retval, sets = 0;
-	zend_long sec, usec = 0;
-	bool secnull;
-	bool usecnull = 1;
-	int set_count, max_set_count = 0;
-
-	ZEND_PARSE_PARAMETERS_START(4, 5)
-		Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
-		Z_PARAM_LONG_OR_NULL(sec, secnull)
-		Z_PARAM_OPTIONAL
-		Z_PARAM_LONG_OR_NULL(usec, usecnull)
-	ZEND_PARSE_PARAMETERS_END();
-
-	FD_ZERO(&rfds);
-	FD_ZERO(&wfds);
-	FD_ZERO(&efds);
-
-	if (r_array != NULL) {
-		set_count = stream_array_to_fd_set(r_array, &rfds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+PHP_FUNCTION(stream_select) {
+    zval *r_array, *w_array, *e_array;
+    struct timeval tv, *tv_p = NULL;
+    fd_bigset rfds, wfds, efds;
+    php_socket_t max_fd = -1;
+    int retval = 0;
+    zend_long sec, usec = 0;
+    bool secnull;
+    bool usecnull = 1;
+
+    ZEND_PARSE_PARAMETERS_START(4, 5)
+        Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
+        Z_PARAM_LONG_OR_NULL(sec, secnull)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG_OR_NULL(usec, usecnull)
+    ZEND_PARSE_PARAMETERS_END();
+
+    // Determine the largest file descriptor
+    if (r_array != NULL) {
+        max_fd = stream_array_get_max_fd(r_array);
+    }
+    if (w_array != NULL) {
+        php_socket_t tmp_max_fd = stream_array_get_max_fd(w_array);
+        if (tmp_max_fd > max_fd) {
+            max_fd = tmp_max_fd;
+        }
+    }
+    if (e_array != NULL) {
+        php_socket_t tmp_max_fd = stream_array_get_max_fd(e_array);
+        if (tmp_max_fd > max_fd) {
+            max_fd = tmp_max_fd;
+        }
+    }
 
-	if (w_array != NULL) {
-		set_count = stream_array_to_fd_set(w_array, &wfds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+    if (max_fd == -1) {
+        zend_value_error("No stream arrays were passed or all arrays were empty");
+        RETURN_THROWS();
+    }
 
-	if (e_array != NULL) {
-		set_count = stream_array_to_fd_set(e_array, &efds, &max_fd);
-		if (set_count > max_set_count)
-			max_set_count = set_count;
-		sets += set_count;
-	}
+    // Allocate fd_bigset based on max_fd
+    FD_BIGSET_ZERO(&rfds, max_fd + 1);
+    FD_BIGSET_ZERO(&wfds, max_fd + 1);
+    FD_BIGSET_ZERO(&efds, max_fd + 1);
```

Comment:
Won't this allocate a ton of memory if I pass a single stream with fd `[65000]` even though I only want to select a single stream?

Reply:
This will allocate 8KiB. This is one of the reasons why the select syscall is not a great API, but in my opinion it's better than the current behavior of stream_select()

Reply:
> Won't this allocate a ton of memory if I pass a single stream with fd `[65000]` even though I only want to select a single stream?

Realisticly, a select on a single fd that high would be very rare. The select call is typically part of an event loop handling hundreds or thousands of tasks, followed by a single select call per iteration. Also before this you could not use select() on fd 65000 at all.